File: ios/Sources/TaigiKeyboard/Lexicon/Database/CustomDictionaryCapacityPolicy.swift
import Foundation
import SQLite3

/// Capacity policy for the custom-dictionary table — two quotas, one per
/// `CustomDictionaryEntry.Origin`.
///
/// The MANUAL quota (`maxEntries`, aligned with Android's `MAX_ENTRIES`) is a
/// hard cap the user runs into; the LEARNED quota (`maxLearnedEntries`, §50)
/// is enforced by eviction so learning never fails. Helpers run inside a
/// serialized `SQLiteConnectionManager.execute { db in ... }` block.
/// Stateless — all methods operate on a caller-provided `OpaquePointer`.
enum CustomDictionaryCapacityPolicy {
    /// Maximum number of MANUAL rows (`origin = 0`).
    // CROSS-PLATFORM INVARIANT — mirrors android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/CustomDictionaryCapacityPolicy.kt (MAX_ENTRIES).
    // Drift causes silent divergence.
    static let maxEntries = 30000

    /// Maximum number of LEARNED rows (`origin = 1`); past it the
    /// fewest-composed, then least recently touched, row goes (ChiaKey's
    /// policy) so a learn never fails.
    // CROSS-PLATFORM INVARIANT — mirrors android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/CustomDictionaryCapacityPolicy.kt (MAX_LEARNED_ENTRIES).
    // Drift causes silent divergence.
    static let maxLearnedEntries = 2000

    /// Drop learned rows past `cap` — side keys first, so the subquery still
    /// resolves against the intact main table. `OFFSET cap` selects exactly
    /// the rows past the cap (none when under it), so no count is needed.
    static func evictLearnedPastCap(db: OpaquePointer, cap: Int = maxLearnedEntries) {
        let pastCap = """
            SELECT id FROM \(CustomDictionarySchema.tableName)
            WHERE origin = 1
            ORDER BY learn_count DESC, updated_at DESC, id
            LIMIT -1 OFFSET ?
        """
        for table in [
            "\(CustomDictionarySchema.searchKeyTableName) WHERE \(CustomDictionarySchema.searchKeyEntryIdColumn) IN (\(pastCap))",
            "\(CustomDictionarySchema.tableName) WHERE id IN (\(pastCap))",
        ] {
            var stmt: OpaquePointer?
            guard sqlite3_prepare_v2(db, "DELETE FROM \(table);", -1, &stmt, nil) == SQLITE_OK else { continue }
            sqlite3_bind_int(stmt, 1, Int32(cap))
            sqlite3_step(stmt)
            sqlite3_finalize(stmt)
        }
    }

    /// True when a row with the given `id` already exists.
    static func entryExists(db: OpaquePointer, id: String) -> Bool {
        var stmt: OpaquePointer?
        defer { sqlite3_finalize(stmt) }
        guard sqlite3_prepare_v2(
            db,
            "SELECT 1 FROM \(CustomDictionarySchema.tableName) WHERE id = ? LIMIT 1;",
            -1, &stmt, nil,
        ) == SQLITE_OK else {
            return false
        }
        stmt.bindText(1, id)
        return sqlite3_step(stmt) == SQLITE_ROW
    }

    /// Current MANUAL row count — learned rows (§50) have their own quota
    /// (`maxLearnedEntries`) and never eat into the user's. Returns 0 on
    /// prepare failure so the caller can treat connection problems as "not
    /// full" — subsequent writes will surface the underlying error.
    static func currentEntryCount(db: OpaquePointer) -> Int {
        (try? sqliteQueryScalarInt(
            db: db,
            "SELECT COUNT(*) FROM \(CustomDictionarySchema.tableName) WHERE origin = 0;",
        )) ?? 0
    }

    /// Throw if inserting would exceed `maxEntries`. Updating an existing
    /// row (same `id`) is not an insert and bypasses the check.
    /// Must run inside the same transaction as the write to avoid TOCTOU.
    static func guardInsertCapacity(db: OpaquePointer, id: String) throws {
        if entryExists(db: db, id: id) {
            return
        }
        guard currentEntryCount(db: db) < maxEntries else {
            throw LexiconError.queryExecutionFailed(
                "Custom dictionary is full (max \(maxEntries) entries)",
            )
        }
    }

    /// Remaining capacity headroom. Negative values clamp to 0.
    static func remainingCapacity(db: OpaquePointer) -> Int {
        max(0, maxEntries - currentEntryCount(db: db))
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/CustomDictionaryCapacityPolicy.kt
package com.siansiansu.taigikeyboard.ime.dictionary

import android.database.sqlite.SQLiteDatabase

/**
 * Capacity policy for the custom-dictionary table. Owns the row-count cap
 * ([MAX_ENTRIES]) and the helpers needed to enforce it inside the same
 * transaction as the write (TOCTOU-safe). Stateless — every method operates on
 * a caller-provided [SQLiteDatabase], mirroring iOS `CustomDictionaryCapacityPolicy`
 * (which takes an `OpaquePointer`). Platform DB-policy code, NOT shared-core.
 *
 * The pure decision helpers ([wouldExceedCap], [remainingCapacity]) are the
 * JVM-testable seam — they take a row count instead of a DB handle, so the
 * boundary can be exercised without inserting 30000 rows.
 */
internal object CustomDictionaryCapacityPolicy {
    // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Lexicon/Database/CustomDictionaryCapacityPolicy.swift (maxEntries).
    // Drift causes silent divergence.
    /** Maximum number of MANUAL rows (`origin = 0`). */
    const val MAX_ENTRIES = 30_000

    /**
     * Maximum number of LEARNED rows (`origin = 1`, §50); past it the
     * fewest-composed, then least recently touched, row goes so a learn
     * never fails.
     */
    // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/Lexicon/Database/CustomDictionaryCapacityPolicy.swift (maxLearnedEntries).
    // Drift causes silent divergence.
    const val MAX_LEARNED_ENTRIES = 2_000

    private const val TABLE_NAME = "custom_dictionary"

    /** MANUAL row count — learned rows have their own quota. 0 on query failure ("not full"). */
    fun currentEntryCount(db: SQLiteDatabase): Int =
        db.rawQuery("SELECT COUNT(*) FROM $TABLE_NAME WHERE origin = 0", null).use {
            if (it.moveToFirst()) it.getInt(0) else 0
        }

    /** True when a MANUAL row with [id] exists — adopting a learned row is a new manual row for the quota. */
    fun entryExists(
        db: SQLiteDatabase,
        id: String,
    ): Boolean =
        db.rawQuery("SELECT 1 FROM $TABLE_NAME WHERE id = ? AND origin = 0 LIMIT 1", arrayOf(id)).use {
            it.moveToFirst()
        }

    /**
     * Drop learned rows past [cap], never [keptId] (the row the caller just
     * wrote). Side keys first, so the subquery still resolves against the
     * intact main table; `OFFSET cap - 1` selects exactly the rows past the
     * cap once the kept row is set aside. Caller holds the transaction.
     */
    // CROSS-PLATFORM INVARIANT — mirrors ios CustomDictionaryCapacityPolicy.evictLearnedPastCap. Drift causes silent divergence.
    fun evictLearnedPastCap(
        db: SQLiteDatabase,
        keptId: String,
        cap: Int = MAX_LEARNED_ENTRIES,
    ) {
        val args = arrayOf<Any>(keptId, (cap - 1).coerceAtLeast(0))
        db.execSQL("DELETE FROM custom_search_key WHERE entry_id IN ($LEARNED_PAST_CAP_SQL)", args)
        db.execSQL("DELETE FROM $TABLE_NAME WHERE id IN ($LEARNED_PAST_CAP_SQL)", args)
    }

    /** The learned rows past the cap, excluding one kept id — `internal` for the JVM SQL test. */
    internal const val LEARNED_PAST_CAP_SQL =
        "SELECT id FROM $TABLE_NAME WHERE origin = 1 AND id <> ? " +
            "ORDER BY learn_count DESC, updated_at DESC, id LIMIT -1 OFFSET ?"

    /**
     * Pure decision: would inserting push past the cap? Updating an existing row
     * ([isExistingRow] = true) is not an insert and never exceeds.
     */
    fun wouldExceedCap(
        currentCount: Int,
        isExistingRow: Boolean,
    ): Boolean = !isExistingRow && currentCount >= MAX_ENTRIES

    /** Remaining capacity headroom. Negative values clamp to 0. */
    fun remainingCapacity(currentCount: Int): Int = (MAX_ENTRIES - currentCount).coerceAtLeast(0)

    /**
     * Throw [CustomDictionaryFullException] if inserting a NEW [id] would exceed
     * [MAX_ENTRIES]. Updating an existing row bypasses the check. Must run in the
     * same transaction as the write to avoid a TOCTOU race with concurrent writers.
     */
    fun guardInsertCapacity(
        db: SQLiteDatabase,
        id: String,
    ) {
        if (wouldExceedCap(currentEntryCount(db), entryExists(db, id))) {
            throw CustomDictionaryFullException(MAX_ENTRIES)
        }
    }
}

/** Thrown when a custom-dictionary insert would exceed the row cap. */
class CustomDictionaryFullException(
    maxEntries: Int,
) : Exception("Custom dictionary is full (max $maxEntries entries)")
